File: android/app/src/main/java/com/bwqr/mavinote/viewmodels/AuthViewModel.kt
package com.bwqr.mavinote.viewmodels

import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class AuthViewModel {
    suspend fun login(email: String, password: String) {
        return suspendCancellableCoroutine { cont ->
            val onceId = Runtime.instance.startOnce(Once(
                onNext = { cont.resume(Unit) },
                onError = { cont.resumeWithException(it) },
                onStart = { _login(it, email, password) }
            ))

            cont.invokeOnCancellation {
                Runtime.instance.abortOnce(onceId)
            }
        }
    }

    private external fun _login(onceId: Int, email: String, password: String): Long
}

File: ios/MaviNote/ViewModels/AuthViewModel.swift
class AuthViewModel {
    func login(_ email: String, _ password: String) async throws -> () {
        return try await withCheckedThrowingContinuation { continuation in
            Runtime.instance().startOnce(Once(
                onNext: { deserializer in continuation.resume(returning: ())},
                onError: { continuation.resume(throwing: $0)},
                onStart: { reax_auth_login($0, email, password)}
            ))
        }
    }
}
